databricks/sdk/mixins/workspace.py

```diff
@@ -1,14 +1,128 @@
-from typing import BinaryIO, Iterator, Optional
+import logging
+from concurrent.futures import ThreadPoolExecutor
+from datetime import datetime
+from queue import Queue
+from threading import Condition
+from typing import BinaryIO, Iterable, Iterator, Optional
 
 from ..core import DatabricksError
 from ..service.workspace import (ExportFormat, ImportFormat, Language,
                                  ObjectInfo, ObjectType, WorkspaceAPI)
 
+_PARALLEL_RECURSIVE_LISTING_HINT_THRESHOLD = 100
+
 
 def _fqcn(x: any) -> str:
     return f'{x.__module__}.{x.__name__}'
 
 
+_LOG = logging.getLogger(__name__)
+
+
+class _ParallelRecursiveListing(Iterable[ObjectInfo]):
+
+    def __init__(self, path, listing, threads, yield_folders, notebooks_modified_after, max_depth):
+        self._path = path
+        self._listing = listing
+        self._threads = threads
+        self._max_depth = max_depth
+        self._yield_folders = yield_folders
+        self._notebooks_modified_after = notebooks_modified_after
+        self._scans = 0
+        self._in_progress = 0
+        # work queue is seeded from main thread and written/consumed from worker threads
+        self._work = Queue(self._threads)
+        # results queue is written from worker threads and consumed from the main thread
+        self._results = Queue(1)
+        # running condition guards self._in_progress, read from main/worker/reporter threads
+        # and modified by worker threads
+        self._cond = Condition()
+        self._reporter_interval = 60
+        self._reporter_cond = Condition()
+
+    def _enter_folder(self, path: str):
+        with self._cond:
+            _LOG.debug(f"Entering folder: {path}")
+            self._in_progress += 1
+        self._work.put(path)
+
+    def _leave_folder(self, path: str):
+        self._work.task_done()
+        with self._cond:
+            self._in_progress -= 1
+            self._scans += 1
+            _LOG.debug(f"Leaving folder: {path}")
+            if self._in_progress != 0:
+                return
+            self._cond.notify_all()
+            _LOG.debug("Sending poison pills to other workers")
+            for _ in range(0, self._threads - 1):
+                self._work.put(None)
+            with self._reporter_cond:
+                self._reporter_cond.notify()
+
+    def _is_running(self):
+        with self._cond:
+            return self._in_progress > 0
+
+    def _reporter(self):
```

Comment:
The reporter is a bit weird. Given that this is a generator and users may not list all of the items "right away", this would just continue to print logs until the generator is consumed. Alternatively, this could be provided as a parameter to the parallel list operation, allowing users to provide a callback that is periodically called with statistics like this if they desire, then they can control when/whether they want to log or do something else.